### ISSUE TITLE
Problem with nested gradient calculation

### ISSUE THREAD

**Issue by jlmaccal** [pytorch/pytorch#47870]
## 🐛 Bug

I have a loss function that requires taking the dot product between the gradient of network and a vector. I give two seemingly identical formulations below. One formulation calculates the correct gradients with respect to network parameters and can minimize the loss to zero. The other formulation gives incorrect gradients and does not reduce the loss.

## To Reproduce

This version reduces the loss to zero:
```python
import torch
import torch.nn as nn
from torch import autograd

net = nn.Sequential(nn.Linear(2, 100), nn.ReLU(), nn.Linear(100, 1))

x = torch.randn(2, 2)
dx = torch.randn(2, 2)
y = torch.tensor([7.0, -12])

optim = torch.optim.Adam(net.parameters(), 1e-3)


for i in range(2000):
    optim.zero_grad()

    q = autograd.Variable(x, requires_grad=True)

    output = net(q)
    grad_output = torch.ones_like(output)

    grads = autograd.grad(
        inputs=q,
        outputs=output,
        grad_outputs=grad_output,
        create_graph=True,
        retain_graph=True,
    )
    grads = grads[0]

    results = []
    for j in range(2):
        results.append(torch.dot(grads[j, :], dx[j, :]))
    result = torch.stack(results)

    result = result.reshape(-1)

    loss = torch.sum((result - y) ** 2)
    loss.backward()
    print(i, loss.item())
    optim.step()
```

This version, where the gradient calculation is moved to a function, does not reduce the loss:
```python
import torch
import torch.nn as nn
from torch import autograd

net = nn.Sequential(nn.Linear(2, 100), nn.ReLU(), nn.Linear(100, 1))

x = torch.randn(2, 2)
dx = torch.randn(2, 2)
y = torch.tensor([7.0, -12])

optim = torch.optim.Adam(net.parameters(), 1e-3)


def calc_grad(x):
    x = autograd.Variable(x, requires_grad=True)
    output = net(x)
    grad_output = torch.zeros_like(output)
    grads = autograd.grad(
        inputs=x,
        outputs=output,
        grad_outputs=grad_output,
        create_graph=True,
        retain_graph=True,
    )
    return grads[0]


for i in range(2000):
    optim.zero_grad()

    grads = calc_grad(x)

    results = []
    for j in range(2):
        results.append(torch.dot(grads[j, :], dx[j, :]))
    result = torch.stack(results)

    result = result.reshape(-1)

    loss = torch.sum((result - y) ** 2)
    loss.backward()
    print(i, loss.item())
    optim.step()
```

## Expected behavior

Both versions of the code should reduce the loss to zero.

## Environment

```
Collecting environment information...
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Mac OSX 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 11.0.0 (clang-1100.0.33.16)
CMake version: version 3.16.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[conda] _pytorch_select           0.1                       cpu_0
[conda] blas                      1.0                         mkl
[conda] libmklml                  2019.0.5                      0
[conda] mkl                       2019.4                      233
[conda] mkl-service               2.3.0            py38hfbe908c_0
[conda] mkl_fft                   1.1.0            py38hc64f4ea_0
[conda] mkl_random                1.1.1            py38h959d312_0
[conda] numpy                     1.19.1           py38h3b9f5b6_0
[conda] numpy-base                1.19.1           py38hcfb5961_0
[conda] pytorch                   1.6.0           cpu_py38hd70000b_0
```

cc @ezyang @gchanan @zou3519 @bdhirsh @albanD @gqchen @pearu @nikitaved

**Comment by ezyang** [comment]
This is indeed pretty strange. I'm guessing it has to do with some variables becoming dead when you factor it out into a variable. But that should result in an error...

**Comment by albanD** [comment]
It took me a while to see it but in your second version you do `grad_output = torch.zeros_like(output)`. With `zeros_like` instead of `ones_like`. So all your gradients will be 0.

Changing that back makes it work just fine :)

**Comment by jlmaccal** [comment]
@albanD Yes, my apologies. This looks like a stupid error on my part.
